framework/src/main/java/org/checkerframework/framework/util/dependenttypes/DependentTypesHelper.java

```diff
@@ -1094,32 +1127,11 @@ private boolean hasDependentType(AnnotatedTypeMirror atm) {
         if (atm == null) {
             return false;
         }
+        // This is a test about the type system.
         if (!hasDependentAnnotations()) {
             return false;
         }
-        boolean b =
-                new SimpleAnnotatedTypeScanner<>(
-                                (type, p) ->
-                                        type.getAnnotations().stream()
-                                                .anyMatch(this::isExpressionAnno),
-                                Boolean::logicalOr,
-                                false)
-                        .visit(atm);
-        return b;
-    }
-
-    /**
-     * Returns the elements of the annotation that are Java expressions.
-     *
-     * @param am AnnotationMirror
-     * @return the elements of the annotation that are Java expressions
-     */
-    private List<String> getListOfExpressionElements(AnnotationMirror am) {
-        for (Class<? extends Annotation> clazz : annoToElements.keySet()) {
-            if (factory.areSameByClass(am, clazz)) {
-                return annoToElements.get(clazz);
-            }
-        }
-        return Collections.emptyList();
+        // This is a test about this specific type.
+        return hasDependentTypeScanner.visit(atm);
```

Comment:
I would move the definition of `hasDependentTypeScanner` next to this method, since this is the only place it is used.

This implementation seems costly and it seems to be called a lot.  I suggest creating a cache and looking up values in the cache.  I guess the cache would map from class name as a string to boolean.

Reply:
> This implementation seems costly and it seems to be called a lot. I suggest creating a cache and looking up values in the cache. I guess the cache would map from class name as a string to boolean.

This is scanning an AnnotatedTypeMirror, so the key of such a map would have to be to be the AnnotatedTypeMirror itself.  Calling hashcode on an AnnotatedTypeMirror is just as inefficient as this annotated type scanner.

Or did you mean the call to `isExpressionAnno` in the scanner itself?  In that case, the implementation of is expression uses a map.